controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitTransactionTask.java

```diff
@@ -0,0 +1,174 @@
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.stream.Collectors;
+
+/**
+ * Request handler for processing commit events in request-stream.
+ */
+@Slf4j
+public class CommitTransactionTask implements StreamTask<CommitEvent> {
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final StreamMetadataStore streamMetadataStore;
+    private final ScheduledExecutorService executor;
+    private final BlockingQueue<CommitEvent> processedEvents;
+
+    @VisibleForTesting
+    public CommitTransactionTask(final StreamMetadataStore streamMetadataStore,
+                                final StreamMetadataTasks streamMetadataTasks,
+                                final ScheduledExecutorService executor,
+                                final BlockingQueue<CommitEvent> queue) {
+        this.streamMetadataStore = streamMetadataStore;
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.executor = executor;
+        this.processedEvents = queue;
+    }
+
+    public CommitTransactionTask(final StreamMetadataTasks streamMetadataTasks,
+                                 final StreamMetadataStore streamMetadataStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetadataStore);
+        Preconditions.checkNotNull(streamMetadataTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.streamMetadataStore = streamMetadataStore;
+        this.executor = executor;
+        this.processedEvents = null;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(CommitEvent event) {
+        // 1. get epoch.. check if the epoch matches the epoch in commit event.. if true, we can proceed.. else throw OperationNotAllowedException
+        // 2. check if txn-commit-node exists.
+        //      if node exists and has same epoch, process the node.
+        // 3. if node doesnt exist, collect all committing txn in the epoch and create txn-commit-node.
+        // 4. loop over the transactions and commit one transaction at a time until all such transactions are committed.
+        // 5. delete txn-commit-node
+        // 6. try complete scale.
+        // Note: Once scale is completed, all events for this epoch will become a no-op as all transactions would have been committed already.
+        String scope = event.getScope();
+        String stream = event.getStream();
+        int epoch = event.getEpoch();
+        OperationContext context = streamMetadataStore.createContext(scope, stream);
+        log.debug("Attempting to commit available transactions on epoch {} on stream {}/{}", event.getEpoch(), event.getScope(), event.getStream());
+
+        return streamMetadataStore.getActiveEpoch(scope, stream, context, false, executor).thenComposeAsync(pair -> {
```

Comment:
Pass the executor to this method. Otherwise this will run on the ForkJoinPool.

Reply:
replaced thenComposeAsync with thenCompose.
ThenComposeAsync came because of copy pasting that part of the code from the original class (CommitEventProcessor.java) which has been replaced by this new class.